### PR TITLE
detect bad role type in is_window

### PR DESCRIPTION
**Pull request by asmagill**
Should address #1029

### CODE REVIEW

File: extensions/uielement/internal.m [hunk 1]
```diff
@@ -42,15 +42,20 @@ static bool is_window(AXUIElementRef element, NSString* role) {
...
+        return YES;
+      } else {
+        return NO;
+      }
   } else {
-    return NO;
+      [LuaSkin logBreadcrumb:[NSString stringWithFormat:@"%s:is_window AXRole is not a string type", watcherTag]] ;
+      return NO;
   }
 }
```
Comment (sprig) [inline extensions/uielement/internal.m]: I think that you also need to amend the push_element function:

Changing
``` lua
// Use the Role of the element to decide which type of object to create: window, app, or plain uielement.
// Retains a copy of the element, if necessary.
static void push_element(lua_State* L, AXUIElementRef element) {
    NSString* role = get_prop(element, NSAccessibilityRoleAttribute, @"");

    if (is_window(element, role)) {
        new_window(L, (AXUIElementRef)CFRetain(element));
    } else if ([role isEqualToString: (NSString*)kAXApplicationRole]) {
        pid_t pid;
        AXUIElementGetPid(element, &pid);
        if (!new_application(L, pid)) {
            lua_pushnil(L);
        }
    } else {
        new_uielement(L, (AXUIElementRef)CFRetain(element));
    }
}
```
to
``` lua
// Use the Role of the element to decide which type of object to create: window, app, or plain uielement.
// Retains a copy of the element, if necessary.
static void push_element(lua_State* L, AXUIElementRef element) {
    NSString* role = get_prop(element, NSAccessibilityRoleAttribute, @"");

    if (is_window(element, role)) {
        new_window(L, (AXUIElementRef)CFRetain(element));
    } else if ([role isKindOfClass:[NSString class]] && [role isEqualToString: (NSString*)kAXApplicationRole]) {
        pid_t pid;
        AXUIElementGetPid(element, &pid);
        if (!new_application(L, pid)) {
            lua_pushnil(L);
        }
    } else {
        new_uielement(L, (AXUIElementRef)CFRetain(element));
    }
}
```
(single line change)

**Comment by asmagill** [comment]
That's probably pretty close to what I will add, though I want to read up a little more and see if we can add some logging to identify the specific applications/windows with this problem... it may be that we should be filtering them another way or that it might be worth submitting a bug report to them about not using Accessibility correctly -- but first we have to identify them!

Keep me posted on how it works for you.

**Comment by asmagill** [comment]
@sprig, this has additional logging that attempts to identify the role type and the application it's from... I'd appreciate it if you could test with this and, if it seems to fix the crashes, execute the following in the Hammerspoon console every so often and report what it displays: `hs.logger.printHistory(nil,nil,"uielement")`.

If it does continue to crash, your crash logs have been useful so far, but you might also check the system log for lines with "hs.uielement" in the message portion and copy these into here as well... by using the logVerbose method, it should put the log messages in both the system log *and* the Hammerspoon console log.

**Comment by sprig** [comment]
@asmagill Thanks, will do

**Comment by sprig** [comment]
Quick update:

Newly patched HS has been running since the last comment. No crashes so far.

``` lua
> hs.logger.printHistory(nil,nil,"uielement")


> 
```
(Nothing to report)

**Comment by sprig** [comment]
Still no crashes and nothing from the logger

**Pull request merged**